src/main/java/bio/terra/cli/context/Resource.java

```diff
@@ -0,0 +1,295 @@
+package bio.terra.cli.context;
+
+import bio.terra.cli.command.exception.UserActionableException;
+import bio.terra.cli.context.resources.AiNotebook;
+import bio.terra.cli.context.resources.BqDataset;
+import bio.terra.cli.context.resources.GcsBucket;
+import bio.terra.cli.context.utils.Printer;
+import bio.terra.cli.service.WorkspaceManagerService;
+import bio.terra.workspace.model.AccessScope;
+import bio.terra.workspace.model.CloningInstructionsEnum;
+import bio.terra.workspace.model.ControlledResourceIamRole;
+import bio.terra.workspace.model.ControlledResourceMetadata;
+import bio.terra.workspace.model.ManagedBy;
+import bio.terra.workspace.model.PrivateResourceUser;
+import bio.terra.workspace.model.ResourceDescription;
+import bio.terra.workspace.model.ResourceMetadata;
+import bio.terra.workspace.model.ResourceType;
+import bio.terra.workspace.model.StewardshipType;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;
+import java.io.PrintStream;
+import java.util.List;
+import java.util.UUID;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * POJO class that represents a workspace resource. This class is serialized to disk as part of the
+ * global context. It is also used as the user-facing JSON output for commands that return a
+ * resource.
+ */
+@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, include = JsonTypeInfo.As.PROPERTY, property = "@class")
+public abstract class Resource {
+  // all resources
+  public final UUID resourceId;
```

Comment:
I think all the fields in these classes either have getters or are `final`. Agreed on not exposing mutable properties. But what are the reasons to prefer a getter over a `final` property?

Reply:
go/java-practices/visibility#public_fields

Reply:
OK that looks like just a style preference. If there's a technical reason, I'll change it, but otherwise I think it makes these classes more cluttered.